### PR TITLE
Feat upload

### LINKED ISSUE #14: Support for secondary directory to get helm warehouse
Thanks for creating an issue! Please fill out this form so we can be
sure to have all the information we need, and to minimize back and forth.

* **What are you trying to do?**
Support for secondary directory to get helm warehouse
* **What feature or behavior is this required for?**
Support for secondary directory to get helm warehouse
* **How could we solve this issue? (Not knowing is okay!)**

* **Anything else?**

helm repo like this

https://burdenbear.github.io/kube-charts-mirror/

image

In this case, there is a problem with the method of setting the relative path before, and the real helm charts will not be obtained.

### PR DESCRIPTION
**Pull request by loverto**
(brief, plain english overview of your changes here)

This pull request makes the following changes:
* support web page upload helm charts
* support seconds dir helm charts repo proxy

image

It relates to the following issue #s:
* Fixes #14 #13

### CODE REVIEW

File: src/main/java/org/sonatype/repository/helm/internal/metadata/IndexYamlAbsoluteUrlRewriter.java [hunk 1]
```diff
@@ -108,10 +108,10 @@ private Event maybeSetAbsoluteUrlAsRelative(ScalarEvent scalarEvent) throws Malf
       URI uri = new URIBuilder(oldUrl).build();
       if (uri.isAbsolute()) {
         String fileName = uri.getPath();
-        // Strip leading slash
-        if (!fileName.isEmpty()) {
-            fileName = fileName.substring(1);
-        }
+        //remove leading "/" to avoid URI missing, by ref: https://github.com/helm/helm/issues/3878
```
Comment (DarthHater) [inline src/main/java/org/sonatype/repository/helm/internal/metadata/IndexYamlAbsoluteUrlRewriter.java]: Pretty sure this is not necessary, the last change should do the same thing.

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    private final VariableResolverAdapter variableResolverAdapter;
+
+    private final HelmAttributeParser helmPackageParser;
+
+    private final HelmPathUtils helmPathUtils;
+
+    @Inject
+    public HelmUploadHandler(final ContentPermissionChecker contentPermissionChecker,final HelmPathUtils helmPathUtils,final HelmAttributeParser helmPackageParser, @Named("simple") final VariableResolverAdapter variableResolverAdapter, final Set<UploadDefinitionExtension> uploadDefinitionExtensions){
+        super(uploadDefinitionExtensions);
+        this.contentPermissionChecker=contentPermissionChecker;
```
Comment (DarthHater) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Spacing is off here, should be: `this.contentPermissionChecker = contentPermissionChecker;`

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+
+    @Override
+    public UploadResponse handle(Repository repository, ComponentUpload upload) throws IOException {
+        HelmHostedFacet facet = repository.facet(HelmHostedFacet.class);
+        StorageFacet storageFacet = repository.facet(StorageFacet.class);
+
+
+        return TransactionalStoreBlob.operation.withDb(storageFacet.txSupplier()).throwing(IOException.class).call(()->{
+            //获取上传信息
```
Comment (DarthHater) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: I had to use Google Translate here :). The comment looks largely informational, maybe just remove it?

File: src/main/java/org/sonatype/repository/helm/internal/metadata/IndexYamlAbsoluteUrlRewriter.java (hunk 1 above)
Comment (loverto) [reply src/main/java/org/sonatype/repository/helm/internal/metadata/IndexYamlAbsoluteUrlRewriter.java]: The modification of this place is not the same as before, here will solve the problem of a secondary directory, like this chart repo address

https://burdenbear.github.io/kube-charts-mirror

https://kubernetes.oss-cn-hangzhou.aliyuncs.com/charts

Comment (DarthHater) [reply src/main/java/org/sonatype/repository/helm/internal/metadata/IndexYamlAbsoluteUrlRewriter.java]: Didn't catch that at first! Could you make a different PR for this change? We try and keep PRs isolated to one feature if possible, so it's more apparent in the merge what was done.

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
+package org.sonatype.repository.helm;
```
Comment (mpiggott) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: This should have a license header

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+
+    @Override
+    public UploadResponse handle(Repository repository, ComponentUpload upload) throws IOException {
+        HelmHostedFacet facet = repository.facet(HelmHostedFacet.class);
+        StorageFacet storageFacet = repository.facet(StorageFacet.class);
+
+
+        return TransactionalStoreBlob.operation.withDb(storageFacet.txSupplier()).throwing(IOException.class).call(()->{
```
Comment (mpiggott) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: I think this is following my bad pattern, the temp blob creation and extraction of metadata ought to occur outside of the transaction - e.g. https://github.com/sonatype/nexus-public/blob/master/plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/PyPiUploadHandler.java#L76

File: src/main/java/org/sonatype/repository/helm/internal/proxy/HelmProxyFacetImpl.java [hunk 6]
```diff
@@ -183,6 +183,11 @@ protected Content doCreateOrSaveComponent(final HelmAttributes helmAttributes,
     return helmDataAccess.saveAsset(tx, asset, componentContent, payload);
   }
 
+  /**
+   * 获取资源
```
Comment (mpiggott) [inline src/main/java/org/sonatype/repository/helm/internal/proxy/HelmProxyFacetImpl.java]: Not sure what this means

**Comment by mpiggott** [comment]
There are quite a few code formatting issues we have Eclipse & IntelliJ formatters available - https://github.com/sonatype/codestyle

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            PartPayload payload = upload.getAssetUploads().get(0).getPayload();
+            TempBlob tempBlob = storageFacet.createTempBlob(payload,HASH_ALGORITHMS);
+            HelmAttributes attributesFromInputStream = helmPackageParser.getAttributesFromInputStream(tempBlob.get());
+
+            String name = attributesFromInputStream.getName();
+            String version = attributesFromInputStream.getVersion();
```
Comment (mpiggott) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Probably need validation for name & version here.

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            String extension = "tgz";
+
+            String path = name + "-" + version + "." + extension;
+
+            ensurePermitted(repository.getName(),HelmFormat.NAME,path, Collections.EMPTY_MAP);
+            try {
+                Asset upload1 = facet.upload(path, tempBlob, payload);
+                return new UploadResponse(upload1);
+            }finally {
+            }
```
Comment (mpiggott) [inline src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: What does this `try-finally` block achieve?

File: src/main/java/org/sonatype/repository/helm/internal/proxy/HelmProxyFacetImpl.java (hunk 6 above)
Comment (mpiggott) [reply src/main/java/org/sonatype/repository/helm/internal/proxy/HelmProxyFacetImpl.java]: Actually the changes to this file seem generally superfluous

File: src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java (hunk 4 above)
Comment (DarthHater) [reply src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: +1

**Comment by DarthHater** [comment]
@loverto , if you can, resolve the feedback from @mpiggott (he worked on the team who originally wrote UI Upload), apply our codestyle to the stuff you've done, and split out that path handling into a separate PR (I want to test that a bit more broadly to make sure it works with other stuff). 

Once you get that handled, I'll be more than glad to get this merged in! You are awesome, thanks a ton for doing this!

**Comment by DarthHater** [comment]
Merged this @loverto and released as v0.0.5, let me know if you have any questions, and thanks for the contribution!

**Pull request merged**